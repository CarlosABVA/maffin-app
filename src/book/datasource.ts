import initSqlJs from 'sql.js';
import { DataSource } from 'typeorm';

import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from './entities';

export async function initDB(database: Uint8Array): Promise<DataSource> {
  await initSqlJs({
    // Required to load the wasm binary asynchronously. Of course, you can host it wherever you want
    locateFile: file => `https://sql.js.org/dist/${file}`,
  });

  let datasource: DataSource;

  if (database.length !== 0) {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      database,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
  } else {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
    await createEmptyBook();
  }

  return datasource;
}

async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}
