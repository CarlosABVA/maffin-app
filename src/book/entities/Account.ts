import * as v from 'class-validator';
import {
  BaseEntity,
  Column,
  Entity, JoinColumn,
  ManyToOne,
  Tree,
  TreeParent,
  TreeChildren,
  OneToMany, PrimaryColumn,
} from 'typeorm';

import type Commodity from './Commodity';
import Money from '../Money';
import type Split from './Split';

/**
 * CREATE TABLE accounts (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   name            text(2048) NOT NULL,
 *   account_type    text(2048) NOT NULL,
 *   commodity_guid  CHAR(32) NOT NULL,
 *   commodity_scu   integer NOT NULL,
 *   non_std_scu     integer NOT NULL,
 *   parent_guid     CHAR(32),
 *   code            text(2048),
 *   description     text(2048),
 *   hidden          integer NOT NULL,
 *   placeholder     integer NOT NULL
 * );
 */

@Entity('accounts')
@Tree('nested-set')
export default class Account extends BaseEntity {
  static TYPES = ['ROOT', 'ASSET', 'BANK', 'CASH', 'EQUITY', 'LIABILITY', 'INCOME', 'EXPENSE', 'MUTUAL', 'STOCK'];

  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
  @v.Length(1, 32)
    guid!: string;

  @Column({
    type: 'text',
    length: 2048,
  })
  @v.MaxLength(2048)
    name!: string;

  @Column({
    type: 'text',
    enum: Account.TYPES,
    length: 2048,
    name: 'account_type',
  })
  @v.IsIn(Account.TYPES)
    type!: string;

  @Column({
    type: 'text',
    select: false,
    default: '',
  })
    path!: string;

  @TreeParent()
  @JoinColumn({ name: 'parent_guid' })
  // Seems TreeParent doesn't let us create referencing the guid of the account
  // and needs the whole object...
    parent!: Account;

  @TreeChildren()
    children!: Account[];

  @ManyToOne('Commodity', { eager: true })
  @JoinColumn({ name: 'commodity_guid' })
    fk_commodity!: Commodity | string;

  get commodity(): Commodity {
    return this.fk_commodity as Commodity;
  }

  @OneToMany('Split', (split: Split) => split.fk_account)
    splits!: Split[];

  /**
   * Note this only works if the entity has access to splits
   */
  get total(): Money {
    if (this.type === 'ROOT') {
      // @ts-ignore
      return null;
    }

    const total = this.splits.reduce(
      (acc, split) => acc.add(
        new Money(split.quantityNum / split.quantityDenom, this.commodity.mnemonic),
      ),
      new Money(0, this.commodity.mnemonic),
    );

    if (total.isNegative()) {
      return total.multiply(-1);
    }

    return total;
  }
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Account, 'name', { value: 'Account' });
