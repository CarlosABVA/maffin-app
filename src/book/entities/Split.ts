import * as v from 'class-validator';
import {
  BaseEntity,
  Column,
  Relation,
  Entity, JoinColumn, ManyToOne, PrimaryColumn,
} from 'typeorm';

import type Account from './Account';
import type Transaction from './Transaction';

/**
 * https://wiki.gnucash.org/wiki/SQL#Tables
 *
 * CREATE TABLE splits (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   tx_guid         CHAR(32) NOT NULL,
 *   account_guid    CHAR(32) NOT NULL,
 *   memo            text(2048) NOT NULL,
 *   action          text(2048) NOT NULL,
 *   reconcile_state text(1) NOT NULL,
 *   reconcile_date  timestamp NOT NULL,
 *   value_num       integer NOT NULL,
 *   value_denom     integer NOT NULL,
 *   quantity_num    integer NOT NULL,
 *   quantity_denom  integer NOT NULL,
 *   lot_guid        CHAR(32)
 * );
*/
@Entity('splits')
export default class Split extends BaseEntity {
  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
  @v.Length(1, 32)
    guid!: string;

  @ManyToOne('Transaction')
  @JoinColumn({ name: 'tx_guid' })
    fk_transaction!: Relation<Transaction> | string;

  get transaction(): Transaction {
    return this.fk_transaction as Transaction;
  }

  @ManyToOne('Account')
  @JoinColumn({ name: 'account_guid' })
    fk_account!: Relation<Account> | string;

  get account(): Account {
    return this.fk_account as Account;
  }

  // For stock operations, action is set to Buy
  @PrimaryColumn({
    type: 'varchar',
    length: 2048,
  })
    action?: string = '';

  @Column({
    type: 'integer',
    name: 'value_num',
  })
    valueNum!: number;

  @Column({
    type: 'integer',
    name: 'value_denom',
  })
    valueDenom!: number;

  @Column({
    type: 'integer',
    name: 'quantity_num',
  })
    quantityNum!: number;

  @Column({
    type: 'integer',
    name: 'quantity_denom',
  })
    quantityDenom!: number;

  /**
   * Returns the value of the transaction in the transaction's currency
   * Let's say you have the following split:
   *
   *  <split:value>100000/100</split:value>
   *  <split:quantity>1228501/10000</split:quantity>
   *
   * where the currency of the Transaction is EUR. This means you spent 1000 EUR to buy
   * 122.8501 of whatever the split's account currency is.
   *
   * When the two currencies are the same, both value and quantity are the same.
   *
   * @returns - value in transaction's currency
   */
  get value(): number {
    return this.valueNum / this.valueDenom;
  }

  /**
   * Returns the quantity spent in the currency of the split's account.
   * Let's say you have the following split:
   *
   *  <split:value>100000/100</split:value>
   *  <split:quantity>1228501/10000</split:quantity>
   *
   * where the currency of the Transaction is EUR and the symbol of the account
   * is GOOGL. This means that you spent 1000 EUR to purchase 122.8501 GOOGL
   * stocks.
   *
   * When the two currencies are the same, both value and quantity are the same.
   *
   * @returns - quantity spent in purchasing in the split account's currency
   */
  get quantity(): number {
    return this.quantityNum / this.quantityDenom;
  }
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Split, 'name', { value: 'Split' });
