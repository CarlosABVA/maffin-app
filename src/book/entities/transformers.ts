import { DateTime } from 'luxon';
import { ValueTransformer } from 'typeorm';

export class DateTimeTransformer implements ValueTransformer {
  // eslint-disable-next-line class-methods-use-this
  from(value: string | null): DateTime | null {
    // For some reason, when using TreeRepository this is null at some point
    if (!value) {
      return null;
    }
    const result = DateTime.fromSQL(value);
    if (!result.isValid) {
      throw new Error(`invalid date: ${result.invalidExplanation}`);
    }
    return result;
  }

  // eslint-disable-next-line class-methods-use-this
  to(value: DateTime): string {
    const result = value.toSQLDate();
    if (result === null) {
      throw new Error(`invalid date: ${value.invalidExplanation}`);
    }
    return result;
  }
}
