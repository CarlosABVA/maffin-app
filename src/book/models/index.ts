export { default as InvestmentAccount } from './InvestmentAccount';
