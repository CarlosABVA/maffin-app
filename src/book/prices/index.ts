export { default as PriceDB } from './PriceDB';
export { default as PriceDBMap } from './PriceDBMap';
