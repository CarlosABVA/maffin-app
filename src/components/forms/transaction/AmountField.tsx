import React from 'react';
import classNames from 'classnames';
import { FaArrowUp, FaArrowDown } from 'react-icons/fa';

import Money from '@/book/Money';
import type { Account } from '@/book/entities';
import type { SplitFieldData } from './types';

export type AmountFieldProps = {
  fromAccount: Account,
  splits: SplitFieldData[],
};

export default function AmountField(
  { fromAccount, splits }: AmountFieldProps,
): JSX.Element {
  let money = new Money(0, fromAccount.commodity.mnemonic);
  if (
    splits.length === 0
    || splits.find(split => !split.toAccount || split.toAccount.guid === '')
  ) {
    return (
      <span className="block py-2 px-4 rounded-lg bg-gunmetal-800 text-sm">
        {money.format()}
      </span>
    );
  }

  splits.forEach((split) => {
    money = money.add(new Money(split.amount || 0, fromAccount.commodity.mnemonic));
  });
  // Total split is always the negative of the partial splits
  money = money.multiply(-1);

  return (
    <span
      className={classNames(
        'flex gap-1 items-center py-2 px-4 rounded-lg bg-gunmetal-800 text-sm',
        {
          'bg-green-500/20 text-green-300': money.toNumber() > 0,
          'bg-red-500/20 text-red-300': money.toNumber() < 0,
        },
      )}
    >
      {money.toNumber() > 0 && <FaArrowUp />}
      {money.toNumber() < 0 && <FaArrowDown />}
      {money.format()}
    </span>
  );
}
