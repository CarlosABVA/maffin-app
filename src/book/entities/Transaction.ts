import { DateTime } from 'luxon';
import {
  BaseEntity,
  Column, CreateDateColumn, Entity, JoinColumn, ManyToOne, OneToMany,
  PrimaryColumn,
} from 'typeorm';
import * as v from 'class-validator';

import type Commodity from './Commodity';
import type Split from './Split';
import { DateTimeTransformer } from './transformers';

/**
 * https://wiki.gnucash.org/wiki/SQL#Tables
 *
 * CREATE TABLE transactions (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   currency_guid   CHAR(32) NOT NULL,
 *   num             text(2048) NOT NULL,
 *   post_date       timestamp NOT NULL,
 *   enter_date      timestamp NOT NULL,
 *   description     text(2048)
 * );
 */

@Entity('transactions')
export default class Transaction extends BaseEntity {
  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
    guid!: string;

  @ManyToOne('Commodity', { eager: true })
  @JoinColumn({ name: 'currency_guid' })
    fk_currency!: Commodity | string;

  get currency(): Commodity {
    return this.fk_currency as Commodity;
  }

  @Column({
    type: 'date',
    transformer: new DateTimeTransformer(),
    name: 'post_date',
  })
    date!: DateTime;

  @CreateDateColumn({
    name: 'enter_date',
  })
    enterDate?: Date;

  @OneToMany('Split', (split: Split) => split.fk_transaction)
    splits!: Split[];

  @Column({
    type: 'text',
    length: 2048,
    default: '',
  })
  @v.MaxLength(2048)
    description!: string;
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Transaction, 'name', { value: 'Transaction' });
