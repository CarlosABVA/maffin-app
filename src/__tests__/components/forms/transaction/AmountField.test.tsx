import React from 'react';
import {
  render,
  screen,
} from '@testing-library/react';

import AmountField from '@/components/forms/transaction/AmountField';
import { Account, Commodity } from '@/book/entities';

describe('AmountField', () => {
  it('shows 0 when no splits', () => {
    render(
      <AmountField
        fromAccount={
          {
            guid: '',
            type: '',
            path: '',
            commodity: {
              mnemonic: 'EUR',
            } as Commodity,
          } as Account
        }
        splits={[]}
      />,
    );

    const element = screen.getByText(/0\.00.*/);
    expect(element).toMatchSnapshot();
  });

  it('shows 0 when no toAccount in split', () => {
    render(
      <AmountField
        fromAccount={
          {
            guid: '',
            type: '',
            path: '',
            commodity: {
              mnemonic: 'EUR',
            } as Commodity,
          } as Account
        }
        splits={[
          {
            amount: 200,
            toAccount: {
              guid: '',
              type: '',
              path: '',
              commodity: {
                mnemonic: 'EUR',
              } as Commodity,
            } as Account,
          },
        ]}
      />,
    );

    const element = screen.getByText(/0\.00.*/);
    expect(element).toMatchSnapshot();
  });

  it('shows 0 when no amount in split', () => {
    render(
      <AmountField
        fromAccount={
          {
            guid: '',
            type: '',
            path: '',
            commodity: {
              mnemonic: 'EUR',
            } as Commodity,
          } as Account
        }
        splits={[
          {
            amount: NaN,
            toAccount: {
              guid: '',
              type: '',
              path: '',
              commodity: {
                mnemonic: 'EUR',
              } as Commodity,
            } as Account,
          },
        ]}
      />,
    );

    const element = screen.getByText(/0\.00.*/);
    expect(element).toMatchSnapshot();
  });

  it('adds split inverse amounts with negative result', () => {
    render(
      <AmountField
        fromAccount={
          {
            guid: '',
            type: '',
            path: '',
            commodity: {
              mnemonic: 'EUR',
            } as Commodity,
          } as Account
        }
        splits={[
          {
            amount: -100.40,
            toAccount: {
              guid: 'split1',
              type: '',
              path: '',
              commodity: {
                mnemonic: 'EUR',
              } as Commodity,
            } as Account,
          },
          {
            amount: 200.20,
            toAccount: {
              guid: 'split2',
              type: '',
              path: '',
              commodity: {
                mnemonic: 'EUR',
              } as Commodity,
            } as Account,
          },
        ]}
      />,
    );

    const element = screen.getByText(/-99\.80.*/);
    expect(element).toMatchSnapshot();
  });

  it('adds split inverse amounts with positive result', () => {
    render(
      <AmountField
        fromAccount={
          {
            guid: '',
            type: '',
            path: '',
            commodity: {
              mnemonic: 'EUR',
            } as Commodity,
          } as Account
        }
        splits={[
          {
            amount: 100.40,
            toAccount: {
              guid: 'split1',
              type: '',
              path: '',
              commodity: {
                mnemonic: 'EUR',
              } as Commodity,
            } as Account,
          },
          {
            amount: -200.20,
            toAccount: {
              guid: 'split2',
              type: '',
              path: '',
              commodity: {
                mnemonic: 'EUR',
              } as Commodity,
            } as Account,
          },
        ]}
      />,
    );

    const element = screen.getByText(/99\.80.*/);
    expect(element).toMatchSnapshot();
  });
});
