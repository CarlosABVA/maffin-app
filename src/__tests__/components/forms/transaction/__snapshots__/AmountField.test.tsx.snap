// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountField adds split inverse amounts with negative result 1`] = `
<span
  class="flex gap-1 items-center py-2 px-4 rounded-lg bg-gunmetal-800 text-sm bg-red-500/20 text-red-300"
>
  <svg
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 448 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
    />
  </svg>
  -99.80 €
</span>
`;

exports[`AmountField adds split inverse amounts with positive result 1`] = `
<span
  class="flex gap-1 items-center py-2 px-4 rounded-lg bg-gunmetal-800 text-sm bg-green-500/20 text-green-300"
>
  <svg
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 448 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M34.9 289.5l-22.2-22.2c-9.4-9.4-9.4-24.6 0-33.9L207 39c9.4-9.4 24.6-9.4 33.9 0l194.3 194.3c9.4 9.4 9.4 24.6 0 33.9L413 289.4c-9.5 9.5-25 9.3-34.3-.4L264 168.6V456c0 13.3-10.7 24-24 24h-32c-13.3 0-24-10.7-24-24V168.6L69.2 289.1c-9.3 9.8-24.8 10-34.3.4z"
    />
  </svg>
  99.80 €
</span>
`;

exports[`AmountField shows 0 when no amount in split 1`] = `
<span
  class="block py-2 px-4 rounded-lg bg-gunmetal-800 text-sm"
>
  0.00 €
</span>
`;

exports[`AmountField shows 0 when no splits 1`] = `
<span
  class="block py-2 px-4 rounded-lg bg-gunmetal-800 text-sm"
>
  0.00 €
</span>
`;

exports[`AmountField shows 0 when no toAccount in split 1`] = `
<span
  class="block py-2 px-4 rounded-lg bg-gunmetal-800 text-sm"
>
  0.00 €
</span>
`;
